File: renderers/ios/VarnMotion.swift
import UIKit

/// Drives a change over time rather than applying it between two frames.
///
/// The engine sends the state a node settles at and how long it should take to get there, so nothing
/// here decides what moves or how far. A curve arrives as its four control points, which is what
/// `UIViewPropertyAnimator` takes, so a name never has to mean the same thing in three places.
enum VarnMotion {
    struct Timing {
        let duration: TimeInterval
        let delay: TimeInterval
        let first: CGPoint
        let second: CGPoint
    }

    static func timing(_ value: Any?) -> Timing? {
        guard let carried = value as? [String: Any] else {
            return nil
        }

        let duration = TimeInterval(VarnValue.number(carried["duration"]) ?? 0) / 1000
        guard duration > 0 else {
            return nil
        }

        let easing = carried["easing"] as? [Any] ?? []
        let point = { (at: Int) -> CGFloat in
            at < easing.count ? (VarnValue.number(easing[at]) ?? 0) : 0
        }

        return Timing(
            duration: duration,
            delay: TimeInterval(VarnValue.number(carried["delay"]) ?? 0) / 1000,
            first: CGPoint(x: point(0), y: point(1)),
            second: CGPoint(x: point(2), y: point(3))
        )
    }

    static func animate(_ timing: Timing, _ changes: @escaping () -> Void) {
        let animator = UIViewPropertyAnimator(
            duration: timing.duration,
            controlPoint1: timing.first,
            controlPoint2: timing.second,
            animations: changes
        )

        animator.startAnimation(afterDelay: timing.delay)
    }

    /// Puts a node on screen in the state it arrives from and moves it to the one it settles at.
    ///
    /// The arrival is drawn a frame after the node is placed, since a view animated before it has a
    /// superview and a frame has nothing to animate from.
    static func arrive(
        _ view: UIView,
        entering: [String: Any],
        settled: [String: Any],
        type: String,
        timing: Timing
    ) {
        var start = settled

        for (key, value) in entering {
            start[key] = value
        }

        VarnStyle.apply(start, to: view, type: type)

        DispatchQueue.main.async {
            animate(timing) {
                VarnStyle.apply(settled, to: view, type: type)
            }
        }
    }
}

File: renderers/android/src/main/kotlin/dev/varn/gui/VarnMotion.kt
package dev.varn.gui

import android.view.View
import android.view.animation.Interpolator
import android.view.animation.PathInterpolator
import org.json.JSONArray
import org.json.JSONObject

/**
 * Drives a change over time rather than applying it between two frames.
 *
 * The engine sends the state a node settles at and how long it should take to get there, so nothing
 * here decides what moves or how far. A curve arrives as its four control points, which is what
 * `PathInterpolator` takes, so a name never has to mean the same thing in three places.
 */
object VarnMotion {
    class Timing(val duration: Long, val delay: Long, val curve: Interpolator)

    fun timing(value: Any?): Timing? {
        val carried = value as? JSONObject ?: return null
        val duration = carried.optLong("duration", 0)

        if (duration <= 0) {
            return null
        }

        val easing = carried.optJSONArray("easing") ?: JSONArray()
        val point = { at: Int -> easing.optDouble(at, 0.0).toFloat() }

        return Timing(
            duration,
            carried.optLong("delay", 0),
            PathInterpolator(point(0), point(1), point(2), point(3)),
        )
    }

    /**
     * Moves a view to the opacity and the transform a style settles at.
     *
     * Only what does not disturb what the layout worked out is animated, so a node keeps the frame it
     * was given however it is faded, slid or scaled.
     */
    fun animate(view: View, style: JSONObject, timing: Timing) {
        val transform = style.optJSONObject("transform")
        val scale = transform?.optDouble("scaleX", 1.0)?.toFloat() ?: 1f

        view.animate()
            .alpha(style.optDouble("opacity", 1.0).toFloat())
            .translationX((transform?.optDouble("translateX", 0.0)?.toFloat() ?: 0f) * view.resources.displayMetrics.density)
            .translationY((transform?.optDouble("translateY", 0.0)?.toFloat() ?: 0f) * view.resources.displayMetrics.density)
            .scaleX(scale)
            .scaleY(transform?.optDouble("scaleY", 1.0)?.toFloat() ?: 1f)
            .rotation(transform?.optDouble("rotate", 0.0)?.toFloat() ?: 0f)
            .setDuration(timing.duration)
            .setStartDelay(timing.delay)
            .setInterpolator(timing.curve)
            .start()
    }

    /** Puts a node on screen in the state it arrives from and moves it to the one it settles at. */
    fun arrive(view: View, entering: JSONObject, settled: JSONObject, type: String, density: Float, timing: Timing) {
        val start = JSONObject(settled.toString())

        for (key in entering.keys()) {
            start.put(key, entering.get(key))
        }

        VarnStyle.apply(start, view, type, density)
        view.post { animate(view, settled, timing) }
    }
}
